feat: replace logrus with zap

GeorgeMac opened the pull request:
Closes: https://github.com/flipt-io/flipt/issues/964
Supports: https://github.com/flipt-io/flipt/issues/992 (with a little work to add supporting config)

This swaps out logrus for zap. I thought I could whip this up quickly for a comparison to aid the discussion.

Zap supports a lot of different configuration.
In fact, the config itself is JSON serializable and could be added to Flipt's own config.

https://pkg.go.dev/go.uber.org/zap#Config

However, for now, I decided to copy and adjust the default “development” config to best match Flips current flag configuration.

I went through a number of call-sites and adjusted the logging to match zap conventions.
I have favoured using zap fields and non-sugared logger.
Which, will make the output more favourable to structured encoding (like JSON).
Furthermore, it should reduce the logging overhead as it removes a bunch of string interpolation.

Before:
```
 _____ _ _       _
|  ___| (_)_ __ | |_
| |_  | | | '_ \| __|
|  _| | | | |_) | |_
|_|   |_|_| .__/ \__|
          |_|

Version: dev
Commit:
Build Date:
Go Version: go1.19

DEBU[0000] migrations up to date
DEBU[0000] store: "sqlite" enabled                       server=grpc
DEBU[0000] starting grpc server                          server=grpc
DEBU[0001] starting http server                          server=http

API: http://0.0.0.0:8080/api/v1
UI: http://0.0.0.0:8080

DEBU[0014] list flags                                    request= server=grpc
DEBU[0014] list flags                                    response= server=grpc
INFO[0014] finished unary call with code OK              grpc.code=OK grpc.method=ListFlags grpc.service=flipt.Flipt grpc.start_time="2022-09-09T11:27:15+01:00" grpc.time_ms=0.459 peer.address="127.0.0.1:54494" server=grpc span.kind=server system=grpc
DEBU[0022] create flag                                   request="key:\"foo\"  name:\"foo\"  description:\"The foo flag\"" server=grpc
DEBU[0022] create flag                                   response="key:\"foo\"  name:\"foo\"  description:\"The foo flag\"  created_at:{seconds:1662719243  nanos:535598000}  updated_at:{seconds:1662719243  nanos:535598000}" server=grpc
INFO[0022] finished unary call with code OK              grpc.code=OK grpc.method=CreateFlag grpc.service=flipt.Flipt grpc.start_time="2022-09-09T11:27:23+01:00" grpc.time_ms=3.778 peer.address="127.0.0.1:54494" server=grpc span.kind=server system=grpc
DEBU[0022] get flag                                      request="key:\"foo\"" server=grpc
DEBU[0022] get flag                                      response="key:\"foo\"  name:\"foo\"  description:\"The foo flag\"  created_at:{seconds:1662719243  nanos:535598000}  updated_at:{seconds:1662719243  nanos:535598000}" server=grpc
INFO[0022] finished unary call with code OK              grpc.code=OK grpc.method=GetFlag grpc.service=flipt.Flipt grpc.start_time="2022-09-09T11:27:23+01:00" grpc.time_ms=0.719 peer.address="127.0.0.1:54494" server=grpc span.kind=server system=grpc
```

After:
```
 _____ _ _       _
|  ___| (_)_ __ | |_
| |_  | | | '_ \| __|
|  _| | | | |_) | |_
|_|   |_|_| .__/ \__|
          |_|

Version: dev
Commit:
Build Date:
Go Version: go1.19

2022-09-09T11:23:04.720+0100	DEBUG	sql/migrator.go:83	first run, running migrations...	{"server": "grpc"}
2022-09-09T11:23:04.725+0100	DEBUG	sql/migrator.go:90	migrations complete	{"server": "grpc"}
2022-09-09T11:23:04.726+0100	DEBUG	flipt/main.go:410	store enabled	{"server": "grpc", "driver": "sqlite"}
2022-09-09T11:23:04.727+0100	DEBUG	flipt/main.go:506	starting grpc server	{"server": "grpc"}
2022-09-09T11:23:05.717+0100	DEBUG	flipt/main.go:625	starting http server	{"server": "http"}

API: http://0.0.0.0:8080/api/v1
UI: http://0.0.0.0:8080

2022-09-09T11:23:07.065+0100	DEBUG	server/flag.go:22	list flags	{"server": "grpc", "request": ""}
2022-09-09T11:23:07.066+0100	DEBUG	server/flag.go:35	list flags	{"server": "grpc", "response": ""}
2022-09-09T11:23:07.066+0100	INFO	zap/options.go:212	finished unary call with code OK	{"server": "grpc", "grpc.start_time": "2022-09-09T11:23:07+01:00", "system": "grpc", "span.kind": "server", "grpc.service": "flipt.Flipt", "grpc.method": "ListFlags", "peer.address": "127.0.0.1:54039", "grpc.code": "OK", "grpc.time_ms": 0.464}
2022-09-09T11:23:44.116+0100	DEBUG	server/flag.go:41	create flag	{"server": "grpc", "request": "key:\"foo\" name:\"Foo\" description:\"The foo flag\""}
2022-09-09T11:23:44.124+0100	DEBUG	server/flag.go:43	create flag	{"server": "grpc", "response": "key:\"foo\" name:\"Foo\" description:\"The foo flag\" created_at:{seconds:1662719024 nanos:116544000} updated_at:{seconds:1662719024 nanos:116544000}"}
2022-09-09T11:23:44.124+0100	INFO	zap/options.go:212	finished unary call with code OK	{"server": "grpc", "grpc.start_time": "2022-09-09T11:23:44+01:00", "system": "grpc", "span.kind": "server", "grpc.service": "flipt.Flipt", "grpc.method": "CreateFlag", "peer.address": "127.0.0.1:54039", "grpc.code": "OK", "grpc.time_ms": 8.483}
2022-09-09T11:23:44.226+0100	DEBUG	server/flag.go:14	get flag	{"server": "grpc", "request": "key:\"foo\""}
2022-09-09T11:23:44.227+0100	DEBUG	server/flag.go:16	get flag	{"server": "grpc", "response": "key:\"foo\" name:\"Foo\" description:\"The foo flag\" created_at:{seconds:1662719024 nanos:116544000} updated_at:{seconds:1662719024 nanos:116544000}"}
2022-09-09T11:23:44.227+0100	INFO	zap/options.go:212	finished unary call with code OK	{"server": "grpc", "grpc.start_time": "2022-09-09T11:23:44+01:00", "system": "grpc", "span.kind": "server", "grpc.service": "flipt.Flipt", "grpc.method": "GetFlag", "peer.address": "127.0.0.1:54039", "grpc.code": "OK", "grpc.time_ms": 0.814}
```

markphelps commented on cmd/flipt/export.go:
```diff
@@ -55,7 +56,7 @@ func runExport(ctx context.Context) error {
 	// export to file
 	if exportFilename != "" {
-		l.Debugf("exporting to %q", exportFilename)
+		logger.Debug("exporting flags", zap.String("destination_path", exportFilename))
```
```suggestion
		logger.Debug("exporting", zap.String("destination_path", exportFilename))
```

markphelps replied:
since it exports more than just flags

markphelps commented on cmd/flipt/import.go:
```diff
@@ -64,7 +65,7 @@ func runImport(ctx context.Context, args []string) error {
 		f := filepath.Clean(importFilename)
 
-		l.Debugf("importing from %q", f)
+		logger.Debug("importing flags", zap.String("source_path", f))
```
```suggestion
		logger.Debug("importing", zap.String("source_path", f))
```

markphelps replied:
since it imports more than just flags

markphelps commented on cmd/flipt/main.go:
```diff
@@ -88,15 +89,35 @@ var (
+			Level:            zap.NewAtomicLevelAt(zap.InfoLevel),
+			Development:      true,
+			Encoding:         "console",
+			EncoderConfig:    zap.NewDevelopmentEncoderConfig(),
```
What do you think about adding `DisableCaller: true,` here? This prevents the filename/line number from being added to the log output, which I think is only helpful when someone is actually working on Flipt itself, not running it

GeorgeMac replied:
💯 Agreed. I actually have that applied locally. Brings it closed to the original output.

markphelps commented on cmd/flipt/main.go:
```diff
@@ -88,15 +89,35 @@ var (
+			Level:            zap.NewAtomicLevelAt(zap.InfoLevel),
+			Development:      true,
+			Encoding:         "console",
+			EncoderConfig:    zap.NewDevelopmentEncoderConfig(),
```
like you mentioned in the PR description, it might be worth us looking into creating our own `EncoderConfig`, for instance to allow formatting the `timestamps` to use something like RFC3339 instead of the full output like `2022-09-10T09:49:40.178-0400`

re: https://github.com/uber-go/zap/issues/485#issuecomment-820995909

GeorgeMac replied:
+1 on RFC3339

markphelps commented on internal/telemetry/telemetry.go:
```diff
@@ -92,7 +92,7 @@ func (r *Reporter) report(_ context.Context, info info.Flipt, f file) error {
 	} else {
 		t, _ := time.Parse(time.RFC3339, s.LastTimestamp)
-		r.logger.Debugf("last report was: %v ago", time.Since(t))
+		r.logger.Debug("last report", zap.Time("when", t), zap.Duration("ellapsed", time.Since(t)))
```
```suggestion
		r.logger.Debug("last report", zap.Time("when", t), zap.Duration("elapsed", time.Since(t)))
```

GeorgeMac replied:
I've done this so many times. This typo 😂

markphelps commented on server/evaluator.go:
```diff
@@ -127,7 +128,7 @@ func (s *Server) evaluate(ctx context.Context, r *flipt.EvaluationRequest) (*fli
 
 			if match {
-				s.logger.Debugf("constraint: %+v matches", c)
+				s.logger.Debug("constraint matche", zap.Reflect("constraint", c))
```
```suggestion
				s.logger.Debug("constraint matches", zap.Reflect("constraint", c))
```

GeorgeMac commented:
<img width="1218" alt="Screenshot 2022-09-11 at 08 21 22" src="https://user-images.githubusercontent.com/1253326/189516729-b786ba38-6bf8-4b3c-ba69-0d9c4f0a3992.png">

Cheers @markphelps, I adjusted those log lines and introduced a `zapcore.Config` which captures your suggestions.

- Dropped the caller
- Formatted time as RFC3339
- Added a little colour (with a u) to the log level.

Technically, if byte for byte encoding match with logrus was required, a `zapcore.Encoder` _could_ be implemented.
However, I sense from reading the current `console` implementation; doing it right is possible, but might not be worth the investment.

GeorgeMac commented:
Also, just for context, I also tried out installing zap as the internal gRPC logger implementation.

It works: `grpc_zap.ReplaceGrpcLoggerV2WithVerbosity(logger, 0)` This is all that is needed.

However, it does bring a lot more chatter into the logs.
```
2022-09-11T08:46:25+01:00	INFO	[core][Channel #1] Channel Connectivity change to IDLE	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core]pickfirstBalancer: UpdateSubConnState: 0x140003cc150, {CONNECTING <nil>}	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core][Channel #1] Channel Connectivity change to CONNECTING	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core][Channel #1 SubChannel #2] Subchannel picks a new address "0.0.0.0:9000" to connect	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core]CPU time info is unavailable on non-linux environments.	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core][Channel #1 SubChannel #2] Subchannel Connectivity change to READY	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core]pickfirstBalancer: UpdateSubConnState: 0x140003cc150, {READY <nil>}	{"server": "grpc", "system": "grpc", "grpc_log": true}
2022-09-11T08:46:25+01:00	INFO	[core][Channel #1] Channel Connectivity change to READY	{"server": "grpc", "system": "grpc", "grpc_log": true}
```

We could make setting this configurable. I remember that there was a common gRPC environment variable for enabling logs. We could piggyback on that.

I also tried out passing the option to adjust log-level based on gRPC response codes.
The default implementation treats `OK` as `DEBUG`. Which, I can see the case for.
To avoid cluttering your logs with just business as usual, successful responses:

```
grpc_zap.UnaryServerInterceptor(logger, grpc_zap.WithLevels(grpc_zap.DefaultClientCodeToLevel)),

// ...

2022-09-11T08:49:28+01:00	DEBUG	finished unary call with code OK	{"server": "grpc", "grpc.start_time": "2022-09-11T08:49:28+01:00", "system": "grpc", "span.kind": "server", "grpc.service": "flipt.Flipt", "grpc.method": "ListFlags", "peer.address": "127.0.0.1:49820", "grpc.code": "OK", "grpc.time_ms": 1.806}
```

However, it appears that today Flipt treats gRPC responses as `INFO` and I didn't want to change that for now.

markphelps commented:
Yeah I agree that we should change:

> grpc_zap.UnaryServerInterceptor(logger, grpc_zap.WithLevels(grpc_zap.DefaultClientCodeToLevel)),

But perhaps in the next release, potentially when we also add the JSON log output option? It makes sense to me that OK responses would be logged as DEBUG instead of INFO (as they are today).

I also agree that the change:

> grpc_zap.ReplaceGrpcLoggerV2WithVerbosity(logger, 0) This is all that is needed.

Should perhaps piggyback off the env var used by GRPC, (I remember it having a weird name?). Idk how much extra value it would be to add another config for Flipt to control this value, it seems to me that using the decided upon env var by the gRPC community would be good enough?

markphelps reviewed (approved):
Amazing! Thank you @GeorgeMac for doing this, it was much needed IMO
